File: server/server-game/src/main/java/top/nipuru/prushka/game/util/font/resolver/BackgroundResolver.java
package top.nipuru.prushka.game.util.font.resolver;

import top.nipuru.prushka.game.util.font.FontRepository;
import top.nipuru.prushka.game.util.font.LengthyComponent;
import top.nipuru.prushka.game.util.font.util.Util;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import net.kyori.adventure.text.format.Style;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.minimessage.tag.Modifying;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import net.kyori.adventure.text.minimessage.tree.Node;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public class BackgroundResolver {

    private static final TextColor OFFSET = TextColor.fromHexString("#fffefd");

    private final String tag;
    private final SplitResolver splitResolver;
    private final FontRepository fontRepository;
    private final LengthyComponent prefix;
    private final LengthyComponent suffix;
    private final List<LengthyComponent> middle;
    private final boolean centre;
    private final boolean covered;
    private final Style style;


    public BackgroundResolver(String tag, SplitResolver splitResolver, FontRepository fontRepository, LengthyComponent prefix, LengthyComponent suffix, List<LengthyComponent> middle, boolean centre, boolean covered, Style style) {
        this.tag = tag;
        this.splitResolver = splitResolver;
        this.fontRepository = fontRepository;
        this.prefix = prefix;
        this.suffix = suffix;
        this.middle = middle.stream().sorted(Comparator.comparing(LengthyComponent::getLength).reversed()).toList();
        this.centre = centre;
        this.covered = covered;
        this.style = style;
    }

    public TagResolver resolver() {
        return TagResolver.resolver(tag, new ExtensionBGTag());
    }

    private class ExtensionBGTag implements Modifying {

        private final int rootNode = 0;
        private String font;
        private boolean bold;


        @Override
        public void visit(@NotNull Node current, int depth) {
            if (depth != rootNode) return;
            // 找到 样式节点 获取根节点样式类型
            font = Util.getFont(current);
            bold = Util.isBold(current);
        }

        @Override
        public Component apply(@NotNull Component current, int depth) {
            if (depth != rootNode) return Component.empty();
            TextComponent.Builder head = Component.text();
            TextComponent.Builder bg = Component.text();
            bg.style(style);

            float sw = suffix.getLength() ;
            float width = fontRepository.getTotalWidth(current, bold, font);

            Component split_0 = splitResolver.resolve(-1);

            float fw = 0;
            Component mid = Component.empty();
            for (LengthyComponent text : middle) {
                float lw = width - fw;
                int c = (int) (lw / text.getLength());
                if (c == 0)
                    continue;
                fw += c * text.getLength();
                for (int i = 0; i < c; i++) {
                    mid = mid.append(text.getSymbol())
                            .append(split_0);
                }
            }
            if (fw < width && !covered) {
                LengthyComponent text = middle.get(middle.size() - 1);
                fw += text.getLength();
                mid = mid.append(text.getSymbol())
                        .append(split_0);
            }
            float back = sw + fw;
            if (centre) {
                back += (fw - width) / 2;
            } else if (fw + sw < width) {
                back += width - fw - sw;
            }
            if (style.color() != null && Objects.equals(style.color().value(), OFFSET.value()))
                back -= 1;

            Component split_text = splitResolver.resolve((int) -back - 1);
            Component split_end = splitResolver.resolve((int) Math.ceil(back - width - 1));


            bg.append(prefix.getSymbol())
                    .append(split_0)
                    .append(mid)
                    .append(suffix.getSymbol());

            head.append(bg)
                    .append(split_text)
                    .append(current)
                    .append(split_end);
            return head.build();
        }
    }
}

File: server/server-bukkit/src/main/kotlin/server/bukkit/util/text/resolver/BackgroundResolver.kt
package server.bukkit.util.text.resolver

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.Style
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.minimessage.tag.Modifying
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import net.kyori.adventure.text.minimessage.tree.Node
import server.bukkit.util.text.font.Font
import server.bukkit.util.text.font.FontRepository
import kotlin.math.ceil

class BackgroundResolver(
    private val tag: String,
    private val splitResolver: SplitResolver,
    private val fontRepository: FontRepository,
    private val prefix: WidthComponent,
    private val suffix: WidthComponent,
    middle: List<WidthComponent>,
    private val centre: Boolean,
    private val covered: Boolean,
    private val style: Style
) {

    companion object {
        private val OFFSET = TextColor.fromHexString("#fffefd")!!
    }

    private val middle: List<WidthComponent> = middle.sortedBy { it.width }

    fun resolver(): TagResolver {
        return TagResolver.resolver(tag, ExtensionBGTag())
    }

    private inner class ExtensionBGTag : Modifying {
        private val rootNode = 0
        private var font = Font.DEFAULT
        private var bold = false
        private var italic = false


        override fun visit(current: Node, depth: Int) {
            if (depth != rootNode) return
            // 找到 样式节点 获取根节点样式类型
            font = current.getFont(font)
            bold = current.isBold(bold)
            italic = current.isItalic(italic)
        }

        override fun apply(current: Component, depth: Int): Component {
            if (depth != rootNode) return Component.empty()
            val head = Component.text()
            val bg = Component.text()
            bg.style(style)

            val sw = suffix.width.toFloat()
            val width = fontRepository.getTotalWidth(current, bold, italic, font)

            val split_0 = splitResolver.resolve(-1)

            var fw = 0f
            var mid: Component = Component.empty()
            for (text in middle) {
                val lw = width - fw
                val c = (lw / text.width).toInt()
                if (c == 0) continue
                fw += (c * text.width).toFloat()
                for (i in 0 until c) {
                    mid = mid.append(text.symbol)
                        .append(split_0)
                }
            }
            if (fw < width && !covered) {
                val text = middle[middle.size - 1]
                fw += text.width.toFloat()
                mid = mid.append(text.symbol)
                    .append(split_0)
            }
            var back = sw + fw
            if (centre) {
                back += (fw - width) / 2
            } else if (fw + sw < width) {
                back += width - fw - sw
            }
            if (style.color() != null && style.color()!!.value() == OFFSET.value()) back -= 1f

            val splitText = splitResolver.resolve(-back.toInt() - 1)
            val splitEnd = splitResolver.resolve(ceil((back - width - 1).toDouble()).toInt())


            bg.append(prefix.symbol)
                .append(split_0)
                .append(mid)
                .append(suffix.symbol)

            head.append(bg)
                .append(splitText)
                .append(current)
                .append(splitEnd)
            return head.build()
        }
    }


}
